### ISSUE TITLE
Rudimentary windows support

### ISSUE THREAD

**Issue by btrepp** [purescript/spago#47]
The issue you were already expecting :). I'll open this and document trying to get it running on windows. Maintainers: don't feel compelled to respond yet, I'll try to do as much as I can myself.

Running in cmd.exe, can't even get to running the install commands. Looks like the install script assumes bash, curl/wget and tar to be present.

UPDATE: Looks like npm install would never work, that script downloads binaries that have been prebuilt. Next steps for me getting this to work would be getting stack+haskell and trying to build it myself.

UPDATE2: stack install fails due to win32/turtle version mismatchs. Stack is implying to add ` Win32-2.5.4.1@sha256:e623a1058bd8134ec14d62759f76cac52eee3576711cb2c4981f398f1ec44b85` to extra deps. Will see what we get from that. (looks like https://github.com/Gabriel439/Haskell-Turtle-Library/issues/323 is the upstream issue)

UPDATE3: This built with the above change. Though stack install did have to be ran multiple times, but I do have a spago.exe :)

Trying yarn
> yarn global add purescript-spago
> yarn global v1.9.4
> [1/4] Resolving packages...
> [2/4] Fetching packages...
> [------------------------------------------------------------------------------------------------------------------------------------------------] 0/501(node:11552) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
> info fsevents@1.2.4: The platform "win32" is incompatible with this module.
> info "fsevents@1.2.4" is an optional dependency and failed compatibility check. Excluding it from installation.
> [3/4] Linking dependencies...
> [4/4] Building fresh packages...
> error C:\Users\beau.trepp\AppData\Local\Yarn\Data\global\node_modules\purescript-spago: Command failed.
> Exit code: 1
> Command: .\install
> Arguments:
> Directory: C:\Users\beau.trepp\AppData\Local\Yarn\Data\global\node_modules\purescript-spago
> Output:
> '.\install' is not recognized as an internal or external command,
> operable program or batch file.
> info Visit https://yarnpkg.com/en/docs/cli/global for documentation about this command.

Trying npm
> npm install -g purescript-spago
> C:\Users\beau.trepp\AppData\Roaming\npm\spago -> C:\Users\beau.trepp\AppData\Roaming\npm\node_modules\purescript-spago\spago
> 
> > purescript-spago@0.6.0 install C:\Users\beau.trepp\AppData\Roaming\npm\node_modules\purescript-spago
> > ./install
> 
> '.' is not recognized as an internal or external command,
> operable program or batch file.
> npm ERR! code ELIFECYCLE
> npm ERR! errno 1
> npm ERR! purescript-spago@0.6.0 install: `./install`
> npm ERR! Exit status 1
> npm ERR!
> npm ERR! Failed at the purescript-spago@0.6.0 install script.
> npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     C:\Users\beau.trepp\AppData\Roaming\npm-cache\_logs\2018-12-18T03_17_01_744Z-debug.log

Trying stack install
> Error: While constructing the build plan, the following exceptions were encountered:
> 
> In the dependencies for turtle-1.5.12:
>     Win32-2.6.1.0 from stack configuration does not match >=2.2.0.1 && <2.6  (latest matching
>                   version is 2.5.4.1)
> needed due to spago-0.6.0.0 -> turtle-1.5.12
> 
> Some different approaches to resolving this:
> 
>   * Set 'allow-newer: true' in C:\sr\config.yaml to ignore all version constraints and build anyway.
> 
>   * Consider trying 'stack solver', which uses the cabal-install solver to attempt to find some
>     working build configuration. This can be convenient when dealing with many complicated
>     constraint errors, but results may be unpredictable.
> 
>   * Recommended action: try adding the following to your extra-deps
>     in C:\VisualStudio_Projects\purescript\spago\stack.yaml:
> 
> Win32-2.5.4.1@sha256:e623a1058bd8134ec14d62759f76cac52eee3576711cb2c4981f398f1ec44b85
> 
> Plan construction failed.

**Comment by justinwoo** [comment]
I would like to eventually get AppVeyor or Azure Pipelines to build binaries too, I just don't know how to get them going :smile:

For a long while we had other tools where the Windows builds were uploaded by an actual user...

**Comment by btrepp** [comment]
A CI build would be definitely the way to go, though I believe there's some other issues before it's even usable on windows.

I've managed to get spago built using stack, and I've left notes above. Unfortunately 'spago init' doesn't work. It looks like the dhall file has some encoding issues the windows side struggles with

>  spago init
> spago:
> Error: Invalid input
> 
> (stdin):121:14:
>     |
> 121 | in  upstream â«½ overrides â«½ additions
>     |              ^
> unexpected 'â'
> expecting "!=", "&&", "++", "->", "//", "//\\", "/\", "==", "merge", "||", "", "", "", "", '#', '(', '*', '+', '.', ':', '?', '@', built-in expression, double infinity, double literal, end of input, import, integer literal, label, list literal, natural literal, record type or literal, text literal, union type or literal, or whitespace

Then a spago build

>  spago build
> spago:
>  ./packages.dhall
> 
> Error: Invalid input
> 
> C:\VisualStudio_Projects\purescript\spagotest\packages.dhall:121:14:
>     |
> 121 | in  upstream â«½ overrides â«½ additions
>     |              ^
> unexpected 'â'
> expecting "!=", "&&", "++", "->", "//", "//\\", "/\", "==", "merge", "||", "", "", "", "", '#', '(', '*', '+', '.', ':', '?', '@', built-in expression, double infinity, double literal, end of input, import, integer literal, label, list literal, natural literal, record type or literal, text literal, union type or literal, or whitespace

File is at 
https://gist.github.com/btrepp/c55143fed5f510b7e19097f68ef538ad

**Comment by justinwoo** [comment]
Whoever decided unicode characters were a good default for Dhall formatting was a true sadist

**Comment by f-f** [comment]
@justinwoo there's the `--ascii` flag (and in our case the `Dhall.Pretty.Ascii` encoding in Haskell), so if we can detect that we're running on Windows we can format with ascii.

**Comment by f-f** [comment]
@btrepp thanks for diving into this!

Here are some ideas for a couple of issues there are on Windows:
- the `purs` executable: we could take care of this in the way `psc-package` does, [see here](https://github.com/purescript/psc-package/blob/master/app/Main.hs#L217)
- installing with npm: we could fallback to a `stack install` in case the executable is not precompiled, as [the `purescript` npm package](https://github.com/purescript-contrib/node-purescript#readme)

**Comment by btrepp** [comment]
@f-f No problems.

Thinking about the encoding more, I'm not sure why this is an issue, it may end up being upstream with the dhall library?. Windows should support unicode, Is the template written by the dhall library or is is generated elsewhere?. Might just be that its saving it incorrectly, or some assumption somewhere holds different for *nix and windows.

I think it's finding purs okay. I did get compilation happening, seems more however telling purs to bundle was not working (I remember this being an absolute pain on windows, when I attempted to make my own makefiles for purs). I'll try and look into this more later.

As far as npm install, I think a CI build with the binaries would do long term. Short term, I think it's okay with the process I'm using (self-building). At least until I can verify that spago is in a use-able state on windows. I haven't really done any haskell, and it was simple enough for me to figure out.
Hell I'm a big fan of 'here's the build release on github', it works really well for purs+psc-package, and doesn't couple me to node as much.
